### DISCUSSION TITLE
Is there an async version of nats::jetstream?

### DISCUSSION THREAD

**Discussion by pashinin** [nats-io/nats.rs#443] [Q&A] [answered]
I want to do something like:

```
...
tokio::select! {
    res = consumer.pull() => {
        let msg = res.unwrap();
        tokio::spawn(async move {
            println!("got message {:?}", msg.data);
            msg.ack().unwrap();
        });
    }
    _ = shutdown_signal() => {
        info!("shutting down worker");
    }
}
...
```

But I cant do this because it is not async.

Or how can I catch SIGINT while waiting for a message?

**Comment by derekcollison** [comment]
You can do push based subscribers.. Tyler may have some more complete insight.

**Comment by cortopy** [comment]
just getting familiar with nats and I have a similar question. All consumer factories, like for example [create_or_open](https://docs.rs/nats/0.11.0/nats/jetstream/struct.Consumer.html#method.create_or_open), require a blocking synchronous connection

I'm not sure how to use the jetstream module in an async context

**Comment by derekcollison** [comment]
@spacejam can comment some more here on how to use the client and JetStream with async constructs.

**Comment by spacejam** [comment]
Hey @pashinin, we are currently in the process of getting async (now available again via the `asynk` module of the nats crate) up to parity with the sync client. Until this is released, you can use all of the blocking functionality using the various blocking threadpool implementations that convert blocking operations into async operations. Examples of this are the `blocking` crate, the various tokio blocking operations, etc...

Signal handling is generally best done using the `signal_hook` crate, and you can use one of the async adapters for this in order to join on multiple types of futures at specific suspension points.

**Comment by Jarema** [comment, marked as answer]
as there is a whole new [async client](https://docs.rs/async-nats/latest/async_nats/) , I think this is answered.
